공동체 정보 및 설정 보기 페이지

yyeonzu opened the pull request:
## 🔎 What is this PR?

Resolves #26, Resolves #38 

### Product spec
- 공동체 정보 보기
- 공동체 권한에 따른 공동체 설정 보기 (admin/user)

### Tech spec
별도의 라이브러리 설치 X


### 디자인 (Figma)
admin/user
<img width="373" alt="image" src="https://github.com/user-attachments/assets/88e4e692-b504-4315-8225-81c6b203bd3c">





## ✨ 설명
위의 사진과 같은 공동체 정보&설정 페이지 퍼블리싱입니다.
디자인 가이드에서 확정되지 않은 설정 버튼 클릭 후의 플로우는 _구현되지 않았습니다._

### 공동체 버튼 컴포넌트
`common > CommunityButton.tsx`

공동체 사이드바에서 사용하는 버튼과 설정 버튼의 레이아웃이 동일하여 공통 UI 컴포넌트로 제작했습니다. 
PR을 따로 보낼까 하다가 전부 제가 작업하는 컴포넌트라서 이 PR에 포함합니다.

flex를 이용한 단순한 UI 컴포넌트입니다. 총 5가지 종류의 버튼이 존재해, 타입으로 정의하여 사용했습니다.
buttonConfig를 만들어 props에서 건네는 타입에 따라 컴포넌트에 필요한 value (name, icon)을 불러옵니다. 

사이드바에서 사용하는 화면은 다음 사진과 같습니다.
<img width="287" alt="image" src="https://github.com/user-attachments/assets/4ad530d5-29ab-48e4-ad72-965d526ce4d7">

### 섹션 컨테이너 컴포넌트 
`communityinfosetting > SectionContainer.tsx`

공동체 정보와 공동체 설정을 하나의 섹션으로 설정하고, 텍스트와 기타 컴포넌트를 포함하는 flex UI 컴포넌트를 구현했습니다. 
단순히 gap을 이용하여 작업했습니다. (좌우 패딩의 경우 페이지 컨테이너에서 관리합니다.)

### 공동체 인포 섹션 컴포넌트
`communityinfosetting > CommunityInfoSection.tsx`

프로필 디폴트 이미지가 미정인 관계로, `black200` color을 사용해 background-color만 적용해두었습니다.
이미지를 찾지 못하는 경우 발생하는 alt 역시 색상이나 보여지는 방법이 정해지지 않아 기본값을 따릅니다. 혹시 hover 시에만 alt가 보여야하는 디자인으로 변경된다면 새 컴포넌트가 추가될 예정입니다.

`subtitle`( = '공동체정보')을 제외한 컴포넌트에 보여지는 모든 값은 page에서 불러오는 API를 통해 props로 가져옵니다.

### 공동체 설정 섹션 컴포넌트
`communityinfosetting > CommunitySettingSection.tsx`

사용자 권한에 따라 보여지는 버튼이 달라, 조건부 렌더링을 사용했습니다. 

api 명세에서 권한을 확인하는 부분을 찾지 못해 질문 후 role 상태 관리 함수 제작 예정입니다.

### 공동체 정보 및 설정 페이지

`pages > communityinfosetting > CommunityInfoSettingPage.tsx`

Container라는 styled-component를 만들어 페이지 전체 좌우 패딩 및 header와의 패딩 설정을 진행했습니다. 
이 페이지에서 뷰포트 (100svw)를 이용해 width를 지정했고, 이하의 컴포넌트에서는 %를 사용해 100svw 안에서 동작하도록 설정했습니다.
설정하면서 불편했던 점이 있는데, 논의를 통해 구버전 브라우저 비호환 방지를 위해 svw와 vw를 함께 작성하기로 했는데 빌드 및 실행할 때마다 property overwritten 경고가 나타났습니다. ㅠㅠ 혹시 안불편하셨는지요....

페이지에 존재하는 모든 API 데이터는 이 파일에서 불러올 예정입니다. 아직 API 배포 이전이므로 delay 함수를 util에 만들어 비동기처리를 테스트했습니다.

loading 중이거나 error 발생했을 때 나타나는 페이지에 대해서도 아직 디자인 결정 및 논의가 안된 것 같아 일단 임의로 리턴하였습니다. 추후 loading/error를 전역에서 관리한다면 삭제될 코드입니다.



## 📷 스크린샷 (선택)

## ✅ 테스트 체크리스트

### 공동체 버튼 컴포넌트 테스트 코드

- [x] 공동체 권한 넘기기 버튼을 렌더링한다.
- [x] 공동체 삭제하기 버튼을 렌더링한다.
- [x] 공동체 나가기버튼을 렌더링한다.
- [x] 초대 링크 복사하기버튼을 렌더링한다.
- [x] 상세 정보 설정 버튼을 렌더링한다.
- [x] 버튼을 클릭했을 때 onClick 이벤트를 호출한다.

### 공동체 정보 섹션 컴포넌트 테스트 코드
- [x] 공동체 이미지를 렌더링한다.
- [x] 공동체 이름을 렌더링한다.
- [x] 공동체 멤버 정보를 렌더링한다.
- [x] 공동체 생성 날짜를 렌더링한다.
- [x] 공동체 설명을 렌더링한다.

### 공동체 설정 섹션 컴포넌트 테스트 코드
- [x] 관리자 역할일 때 권한 넘기기 & 삭제하기 옵션을 렌더링한다.
- [x] 일반 사용자 역할일 때 나가기 옵션을 렌더링한다.

## 💡 집중 리뷰 요청

평소 혜린님이 코드를 작성할 때 중요하게 생각하는 부분을 집중적으로 체크해주세요!! 특히 모듈화나 코드를 더 단순히 or 쫌쫌하게 짤 수 있는 방법을 알려주시면 다음 코드 작성때 반영하겠습니다. +_+

또한 제 코드에서 코드리뷰 전 이해해주셨으면 하는 부분을 몇 가지 적어보겠습니다!
**타입과 인터페이스 모두 로컬파일에서 정의했습니다.ㅠㅠ** 생각보다 타입 정의 위치, 선언할 타입 개수 등등 고민하는 데에 시간이 길어져서 과감히 지역에서 사용했고, 따라서 export 및 import도 규칙없이 작성했습니다. 화요일 논의 이후에 한 번에 정리하여 코드 개선하겠습니다. 이 부분은 리뷰 패스 부탁드려요!
테스트 코드의 경우 주로 정상작동과 관련된 케이스만 작성했습니다. 잘못된 props가 전달된 경우, 비정상 이벤트가 발생한 경우와 같은 케이스는 테스트 코드 작성에 익숙해지면 차차 추가하겠습니다. 혹시 제안할만한 테케 or 저와 달리 작성하신 좋은 테케가 있다면 코멘트 달아주시면 감사하겠습니다 (--)(__)

hyerindev commented on src/utils/helpers.ts:
👍 delay를 만들어서 비동기 테스트를 하다니 정말 멋진 아이디어네요!

hyerindev commented on src/utils/helpers.ts:
```diff
@@ -1 +1,9 @@
+    setTimeout(resolve, ms);
+  });
+
+const empty = {};
```
❗ 커밋 시 컴파일러 에러 때문에 임시로 작성한 코드는 제외 부탁드립니다.

yyeonzu replied:
컴파일러 에러 때문은 아니고, 다음과 같은 ESLint 에러 때문에 삽입한건데 이것도 마찬가지로 제외해야할까요?
`ESLint: Prefer default export on a file with single export. (import/prefer-default-export)`

hyerindev replied:
네, 삭제하거나 주석을 달아놓는 게 좋겠어요. 나중에 다른 사람이 코드를 봤을 때 삭제해도 괜찮은 건지 알기가 어려울 것 같아서요.

hyerindev commented on src/utils/helpers.ts:
```diff
@@ -1 +1,9 @@
+
+const empty = {};
+
+export { delay, empty };
```
❗ 그러고보니 이걸 회의에서 논의하는 걸 깜빡했는데 전역 타입이나 util 등은 `export`를 한 번에 하는 게 번거로울 것 같아서요, 저는 global.d.ts 에서 각 타입마다 `declare` 또는 `export` 하였는데 방식을 통일하면 좋겠습니다.

yyeonzu replied:
delay 함수는 타입이 아닌데, 그럼 export를 인라인 방식으로 함수 선언과 동시에 하자는 이야기 맞을까요 ?.? helper 함수의 경우에는 어떤 파일에서 export 하자는 뜻인지 헷갈리네욤...

hyerindev replied:
네 맞아요! 제가 global.d.ts와 util을 둘다 전역 개념으로 생각하고 있어서 그렇게 말했는데 너무 제 관점이었군요 ㅋㅋ

hyerindev replied:
helper가 코드 조각 모음이다보니까 한 파일에 여러 개 util이 작성될 것 같은데, 인라인 export가 아니라 최하단에서 일괄로 export한다면 왔다리갔다리 해야 해서 불편하지 않을까 싶은 생각에 말씀드렸습니다. 공식 규칙으로 지정된 내용은 아니니 회의 후에 다음부터 반영해도 좋아요!

hyerindev commented on src/tests/utils/renderWithProvider.setup.tsx:
💊 나중에 1차 배포 후 코드 리팩토링 시, jest.setup.js에서 전역으로 선언한 `render()`를 사용해주시기 바랍니다.

hyerindev commented on src/utils/helpers.ts:
```diff
@@ -1 +1,9 @@
-export {};
+// ms초 만큼 기다리는 함수
```
💊 이번에 새로 알게되었는데, [JSDoc](https://ilimes.github.io/javascript/post29/) 형식으로 작성하면 코드 작성 중 마우스 hover 시 그 주석이 보여요! 저도 앞으로 이렇게 작성해보려고 합니다.

hyerindev commented on src/components/common/CommunityButton.tsx:
```diff
@@ -0,0 +1,84 @@
+import React from 'react';
```
❗ 파일마다 React를 import하지 않아도 되게끔 설정해놓았습니다 😊

yyeonzu replied:
~~저는,,, 하지 않으면 자꾸 ESLint 에러가 발생해서요 ........ 🥹 혹시 어떤 브랜치 풀 받으면 되는걸까요?~~
원래 React.FC를 통해 컴포넌트의 타입을 지정했는데, 명시적으로 타입 지정하는 방식으로 변경해 해당 ESLint 오류 제거했고, React import도 삭제했습니다.

hyerindev commented on src/components/common/CommunityButton.tsx:
💬 갑자기 떠올렸는데 계정 설정 페이지의 버튼도 같은 UI 컴포넌트를 사용해요. @rwaeng 님께서 이미 개발하셨을텐데 공통 컴포넌트를 잘 활용하지 못하게 되어 아쉽네요. 이 부분도 나중에 1차 배포 후 코드 리팩토링 시 조정하면 좋겠습니다.

hyerindev commented on src/components/common/CommunityButton.tsx:
```diff
@@ -0,0 +1,84 @@
+
+interface CommunityButtonProps {
+  type: CommunityButtonType;
+  onClick?: () => void; // 이벤트가 발생하지 않는 버튼 대비 optional 처리
```
💊 모든 버튼은 이벤트를 발생시키는 것이 목적이므로 이벤트가 발생하지 않는 버튼은 존재하지 않을 것 같은데 필수로 설정하면 어떨까요?

yyeonzu replied:
이 부분은 사실 onClick 이후의 플로우가 아직 결정이 안돼서,,, 테스트 코드를 작성할 때 onClick을 props로 넘겨주기 애매해서 옵셔널 연산자를 사용했습니다. ㅠㅠ onClick 이후 연결되는 컴포넌트 작성하면서 수정하도록 하겠습니다!

hyerindev commented on src/components/common/CommunityButton.tsx:
```diff
@@ -0,0 +1,84 @@
+  return (
+    <ButtonContainer onClick={onClick} data-testid={testId} aria-label={name}>
+      <IconWrapper>
+        <img src={icon} alt={`${name} icon`} />
```
👍 세심하게 alt를 작성하신 점이 좋아요.

hyerindev commented on src/components/common/CommunityButton.tsx:
```diff
@@ -0,0 +1,84 @@
+}) => {
+  const { name, icon } = buttonConfig[type];
+  return (
+    <ButtonContainer onClick={onClick} data-testid={testId} aria-label={name}>
```
❗ `aria-label`을 지정하셨군요! 테스트에서 `getByLabelText()`를 사용하면 `data-testid`를 사용하지 않아도 되어서 코드가 줄어들겠어요. `CommunityButtonProps`에서 옵셔널인 `testId`를 삭제할 수 있고요!

yyeonzu replied:
오 그건 생각을 못했네용,,, 저는 aria-label이나 다른 값은 변경될 수도 있을 것 같아 test용으로 id를 따로 만들어둔거긴 한데 label을 이용하는 것도 좋은 것 같아요. 혹시 이 부분은 다음 코드 작성부터 반영해도 될까요 ?.?

hyerindev replied:
네 좋아요! 😄😊

hyerindev commented on src/components/common/CommunityButton.tsx:
```diff
@@ -0,0 +1,84 @@
+  testId?: string; // 테스트를 하지 않는 버튼 대비 optional 처리
+}
+
+const buttonConfig = {
+  transferAuthority: {
+    name: '공동체 권한 넘기기',
+    icon: biCrown,
+  },
+  deleteCommunity: {
+    name: '공동체 삭제하기',
+    icon: fiTrash,
+  },
+  leaveCommunity: {
+    name: '공동체 나가기',
+    icon: fiLogOut,
+  },
+  copyInvitation: {
+    name: '초대 링크 복사하기',
+    icon: bIUserPlus,
+  },
+  detailInfoSetting: {
+    name: '상세 정보 설정',
+    icon: fiSettings,
+  },
+};
```
💊 버튼의 가짓수를 미리 정해놓는 방식은 추후 다른 페이지에서 이 UI를 사용할 경우의 확장성과 디자인이 수정될 경우의 대응성이 조금 아쉽게 느껴지는데 어떻게 생각하시나요? 저는 공통 컴포넌트에서 UI만 정의하고 확장성을 최대로 추구하는 취향이라서요.

yyeonzu replied:
저 버튼을 Community에서만 사용한다고 생각해 다른 페이지에서 사용 못하게 하려고 일부러 막은 케이스이긴 한데요.!! 
말씀해주신 개인 계정에서도 사용할 수 있을 것 같아서 리팩토링 때 수정해보겠습니다.
공통 컴포넌트에서 UI만 설정하는 방법이 가장 좋다는 것에는 동의하지만, 가끔은 매번 컴포넌트에서 Props를 작성하는 게 귀찮아서 config를 한 번만들어봤습니다. 근데 요것도 은근 편한 것 같아요 ~!! 오히려 컴포넌트를 사용한 곳이 정확히 어디인지 config에서 가시적으로 관리할 수 있는 것 같아서 좋은 것 같기도 하고요.... ㅎㅎ

hyerindev replied:
오 일부러 제한하다니 그 생각은 못했었는데 그 관점도 좋네요! 말씀해주신 것처럼 관리할 수 있어서 통제 범위에서 계산할 수 있는 이점이 있겠어요. 나중에 리팩토링 기대하겠습니다 👍

hyerindev commented on src/tests/common/CommunityButton.test.tsx:
```diff
@@ -0,0 +1,57 @@
+      expect(screen.getByText(name)).toBeInTheDocument();
+      expect(screen.getByAltText(iconAltText)).toBeInTheDocument();
+      expect(screen.getByTestId(type)).toBeInTheDocument();
+    });
```
> 혹시 제안할만한 테케 or 저와 달리 작성하신 좋은 테케가 있다면 코멘트 달아주시면 감사하겠습니다 (--)(__)

💊 버튼 컴포넌트의 모든 커스터마이징 가능한 요소를 테스트하신 점 인상깊었습니다. 여기에 덧붙여 `.toHaveTextContent()`, `.toHaveAttribute()` 등을 사용하면, "버튼을 렌더한다" 외에도 테스트케이스가 다양해질 수 있겠다고 생각이 듭니다.

hyerindev commented on src/components/communityinfosetting/CommunityInfoSection.tsx:
```diff
@@ -0,0 +1,104 @@
+const Description = styled.div`
+  font-family: ${({ theme }) => theme.fonts.caption};
+  color: ${({ theme }) => theme.colors.black200};
+`;
```
❗ 이름 때문에 코드를 두 번 작성하신 것 같은데, `{creationDate}`와 `{description}`로 역할이 표시되어 그 문제는 충분히 해결되었다고 생각해요. 제너럴한 컴포넌트명을 사용하고 코드를 재활용하면 좋겠어요.

hyerindev commented on src/components/communityinfosetting/CommunityInfoSection.tsx:
```diff
@@ -0,0 +1,104 @@
+  width: 100%;
+`;
+
+const Name = styled.div`
```
💊 제가 코드를 작성할 때 중요하게 생각하는 부분 중 하나는 시맨틱 태그를 적극적으로 활용하는 것입니다. 의미 있는 태그를 사용해야 브라우저가 잘 파악하고 스크린리더가 잘 읽어내기 때문입니다. 그래서 div는 최대한 사용하지 않으려 노력하고 있어요. 그런 의미에서 글자를 꾸미는 태그는 `span`을 사용하는 것이 어떤지 추천드립니다 😉😊😀

hyerindev commented on src/components/communityinfosetting/SectionContainer.tsx:
💬 모듈화에 관해서 중요하게 생각하는 부분이 있는지 여쭤봐주셔서 용기내어 작성해보자면... section마다 subtitle이 반복되고 있으므로 저라면 SectionContainer에 그 부분도 함께 넣어 컴포넌트화할 것 같아요. 코드도 재활용할 수 있고, "section의 subtitle"이라는 의미가 더 살아난다고 생각해요. 그런데 역시 이런 부분까지는 취향의 영역인 것 같으니 사담 정도로 여겨주세요.

yyeonzu replied:
맞아요~!! 저도 작성하면서 합칠까 고민했는데, 다른 컴포넌트 만들면서 공통컴포넌트에서 TitleAndFieldContainer를 만들었거든요! ~~아마 그 컴포넌트랑 함께 재사용할 수 있을 것 같아서 추후 리팩토링 때 참고하여 만들어보겠습니다. ㅎㅎ~~

해당 사항 반영해서 리팩토링 완료했습니다!

hyerindev commented on src/tests/communityinfosetting/CommunityInfoSection.test.tsx:
```diff
@@ -0,0 +1,50 @@
+    renderWithProviders(<CommunityInfoSection {...mockData} />);
+    const imageElement = screen.getByAltText(/profile/i);
+    expect(imageElement).toBeInTheDocument();
+    expect(imageElement).toHaveAttribute('src', mockData.imageUrl);
```
👍 src 속성을 갖고 있는지 테스트하는 점이 좋아요.

hyerindev commented on src/tests/communityinfosetting/CommunitySettingSection.test.tsx:
```diff
@@ -0,0 +1,46 @@
+    expect(
+      screen.queryByRole('button', { name: '공동체 삭제하기' }),
+    ).not.toBeInTheDocument();
+    // id 기반 검증
```
❓ role 기반, id 기반으로 2번 체크하셨는데, 단순히 테스트를 강화하려는 목적인가요? 아니면 구분에 따른 다른 목적이 있나요?

yyeonzu replied:
테스트 강화 목적이 맞습니다! 원래 role 기반으로 작성했다가, name이 변경 되는 경우를 고려하여 id 기반을 포함했습니다. 
테스트 코드 통과 이후 단순한 코드 리팩토링이 발생했을 때 테스트 코드에 영향을 주지 않으려는 목적 & 영향을 줬을 때 다른 것이 아닌 이름 변경 때문이라는 것을 확인하기 위한 목적으로 두 가지 방법을 작성했습니다.
테스트 코드에 익숙해지거나, 리팩토링하면서 시행착오를 겪으면 더 편한 방식으로 정착하지 않을까 하는 기대로.. 일단 두 가지 다 적어봤습니다. ㅎㅎ

hyerindev replied:
오 역시 그런 역할이 있었군요! jest를 처음 쓰다보니까 여러 각도로 체크해서 테스트하는 것이 무엇을 고려한 것인지 배우고 싶었어요.  테스트 코드에 영향을 주지 않도록 하고, 영향을 준다면 name 때문임을 알 수 있게 설계한 점이 좋네요. 테스트 코드 수정을 최소화하는 방법을 잘 생각해보지 못했어서 배워갑니다 👍

hyerindev reviewed (dismissed):
PR을 정말 꼼꼼하고 멋지게 작성하셔서 많이 배웠어요. 코드도 정말 좋아서 들뜬 마음에 깊이 뜯어보았는데 대부분 💊이니 심각하게 받아들이지 말아주세요...! 아래는 PR 본문과 관련된 답변입니다. 코드 리뷰는 코멘트를 확인해주시기 바랍니다. 당장 머지해야 하는 상황이 발생할 경우를 고려하여 Approve로 마칩니다.

## property overwritten 경고

> 설정하면서 불편했던 점이 있는데, 논의를 통해 구버전 브라우저 비호환 방지를 위해 svw와 vw를 함께 작성하기로 했는데 빌드 및 실행할 때마다 property overwritten 경고가 나타났습니다. ㅠㅠ 혹시 안불편하셨는지요....

오 부끄럽게도 그 논의를 까맣게 잊고 있다가 말씀해주셔서 기억해냈습니다.
지난 화요일 회의에서 말씀드렸다시피 글로벌스타일에서 아래와 같이 지정하여 이제 `#root`의 하위 요소에서 `svw`, `svh` 단위를 사용할 필요 없이 부모 요소를 참조하는 `%` 단위로 지정 가능합니다. 나중에 1차 배포 후 코드 리팩토링 시에 수정 부탁드립니다.
```css
body {
  width: 100svw;
  height: 100svh;
}
#root {
  width: 100%;
  height: 100%;
}
```

그런데 제가 그 논의를 깜빡한 탓에 글로벌스타일에서 `vw`, `vh`를 함께 작성하지 않았습니다.. property overwritten 경고에 관해 알아보겠습니다. 알려주셔서 감사합니다.

hyerindev commented:
property overwritten 경고를 해결했습니다. #60 확인 부탁드립니다.

yyeonzu committed:
:recycle: 코드 리뷰 반영 (#26)
- 제너럴 컴포넌트로 명명 변경 및 재사용
- text를 감싸는 컴포넌트에서 div 태그 대신 span 태그 사용

yyeonzu commented:
리뷰 사항 일부 반영하여 커밋했습니다 ~! 또한 남겨주신 댓글에 답변도 남겨놨으니 확인 부탁드립니다. 혹시 확인 후 문제 없다면 Approve 상태로 한 번 더 변경해주시면 감사하겠습니다.

+) 한 가지 추가 변경사항이 있는데, 다른 페이지에서도 사용되어 `CommunityInfoSection.tsx`의 `CardContainer` 이하의 모든 코드가 `CommunityInfoCard` 라는 공통 컴포넌트에서 재사용되도록 수정했습니다.

yyeonzu committed:
:recycle: 반복되는 코드 재사용 컴포넌트 사용 & 명시적 타입 표시 (#26)
- 반복되는 코드 모듈화 진행 (TitleANdFieldContainer 컴포넌트 사용)
- 명시적 타입 표시 & 그로 인해 필요 없어진 import React 삭제

yyeonzu committed:
:recycle: CommunityInfoCard 컴포넌트 생성(#26)
다른 페이지에서도 공동체 정보 보기 카드가 사용되어 공통 컴포넌트로 수정

hyerindev reviewed (approved):
코멘트 반영 수고하셨습니다🤗
남겨주신 코멘트 확인 후 저도 일부 코멘트에 답을 남겼습니다, 확인 부탁드려요!

yyeonzu commented:
코멘트 확인했습니다. 머지 후 리팩토링부터 반영하도록 하겠습니다. 감사합니다~
